text/0064-refer-to-types-in-empty-namespace.md

```diff
@@ -0,0 +1,159 @@
+namespace.
+This is correct behavior which we want to preserve.
+However, currently, the schema author has no way to adjust this attribute
+declaration so that it refers to the `User` type declared in the empty
+namespace.
+With this RFC, the schema author can use the syntax `owner: ::User` to
+explicitly refer to the `User` type declared in the empty namespace.
+
+## Motivation
+
+Users justifiably expect that they can refer to any declared typename (in any
+namespace) from any position in the schema, by writing a fully qualified name.
+The fact that RFC 24 did not provide a way to refer to typenames in the empty
+namespace by using some kind of fully qualified name seems like an oversight,
+and should be corrected for consistency.
+
+The other change proposed in this RFC, implicitly falling back to the empty
+namespace (so that the `Demo::User` declaration is legal in the example above),
+is an ergonomic change to avoid having to use the new syntax in the common case
+where there is actually no typename collision (i.e., where the typename is not
+actually defined in both the active and the empty namespace).
+
+The two changes are theoretically separable and/or severable, but make sense to
+address together, as they both relate to how we reference typenames in the empty
+namespace.
+The implicit-fallback change deals with referencing a typename in the empty
+namespace where there is no collision with a typename in the active namespace.
+The explicit-`::` change deals with referencing a typename in the empty
+namespace where there _is_ a collision with a typename in the active namespace.
+
+This issue has been encountered "in the wild" by multiple people, according to
+a [comment on cedar#579](https://github.com/cedar-policy/cedar/issues/579#issuecomment-2077482637).
+
+## Detailed design
+
+### Change 1: Fully qualified syntax for referring to typenames in the empty namespace
+
+The syntax `::id`, where `id` is an unqualified typename, now explicitly refers
+to the `id` in the empty namespace.
+
+This applies:
+* regardless of if there is an `id` defined in the active namespace or not
+* regardless of if there is an `id` defined in the empty namespace or not (if not, this is an error)
+* regardless of whether `id` is the name of a common type or an entity type, in
+either the empty and/or active namespace
+* in both the human (RFC 24) and JSON schema syntaxes
+
+This does not apply:
+* to primitive and extension types. From the user's perspective, those live in
+the `__cedar` namespace and not in the empty namespace. `::String` refers to an
+entity or common type `String` defined in the empty namespace, while
+`__cedar::String` refers to the builtin primitive type `String`.
+* to "names" like `::A::B` -- that syntax remains an error. `A::B` is already an
+unambiguous, fully-qualified name, and will never have any namespace implicitly
+prepended.
+
+### Change 2: Implicitly falling back on the empty namespace
```

Comment:
I think this solution doesn't solve the fundamental issue where there are type name conflicts between a namespace and the empty space.

Reply:
I think it solves the main problem people encounter. Conflicting type names feels like an edge case, what's more important is just being able to refer to types in the empty namespace.

IMO this change could go in without an RFC

Reply:
@shaobo-he-aws I think Change 1 (not Change 2) is intended to solve that issue?  Or am I misunderstanding?

Reply:
I am worried about conflicts, too.

I am also wondering if we should be looking ahead to schemas that will reference other schemas and leverage mechanisms like `import`. If we had `import`, would we feel the need for `::User`? Is `::User` basically an anti-pattern -- you shouldn't be defining entity types at the top level, and should be using namespaces and `import`?

Reply:
> I am worried about conflicts, too.

Can you say more about what you're worried will happen?